### DISCUSSION TITLE
Override UUID/Acknowledge UUID Change without removing device?

### DISCUSSION THREAD

**Discussion by BlackDragon7402** [krahabb/meross_lan#373] [Q&A]
I just upgraded to the new meross cloud version 5.+ (from previous 4.X)
This release also implemented the check of UUID-Match. Which is obviouls a great feature but I ran into an issue with this when upgrading.
Some time ago I got one of my MTS200 replaced due to an hardware defect. Therefore the UUID of the device changed since the initial setup. When swapping the device no configuration changes on HA/meross_lan where needed since I adjusted the IP to match the old one by using a static IP entry.

My question:
Whats the easiest way to get my device "working" again in Home Assistant? Obviously removing and re-integrating the device again should fix the issue? (Would I only need to delete it and Home Assistant auto-discovers it again via the Meross Cloud?)
Is there an easier way? Maybe an option to acknowlodge the UUID change and therefore HA/meross_lan "overrides" the old UUID with the new one? (Ok this may be a feature request...)
If possible I would like to keep my data/history/configuration, so not remove my device
Thanks for any help

Screenshots displaying my issue (in german sorry :/)
![grafik](https://github.com/krahabb/meross_lan/assets/119086185/9ce86b25-53d0-4144-aaa5-f7186d283151)
![grafik](https://github.com/krahabb/meross_lan/assets/119086185/5ef347d0-b19a-47d4-93e1-2f65c4871b16)

**Comment by krahabb** [comment]
This could be tricky ;)
I'm not sure HA doesn't delete all the history if you remove the configuration so we're only left with 'manual fixing'

All is done by manually editing the /.storage/core.config_entries. Here (it might be very huge - better search for the UUID to quickly get to the point) you should see some:
```     
{
        "entry_id": "6a400d6aea8a0f41ca5cdfb0c703d797",
        "version": 1,
        "minor_version": 1,
        "domain": "meross_lan",
        "title": "mss310r - With wrong UUID",
        "data": {
          "device_id": "#likely the old UUID: this need to be fixed#",
          "payload": {
            "all": {
              "system": {
                "hardware": {
                  "type": "mss310r",
                  "subType": "un",
                  "version": "2.0.0",
                  "chipType": "mt7682",
                  "uuid": "#likely the new UUID - else fix it#",
                  "macAddress": "#also likely the new mac - else fix it#"
                },
                "firmware": {
                  "version": "2.1.4",
                  "compileTime": "2019/11/27 11:48:08 GMT +08:00",
                  "wifiMac": "#whatever#",
                  "innerIp": "#whatever#",
                  "server": "#whatever#",
                  "port": 443,
                  "userId": #whatever#
                },
                "time": {
                  "timestamp": 1704977652,
                  "timezone": "Europe/Rome",
                  "timeRule": [
                    [
                      1698541200,
                      3600,
                      0
                    ],
                    [
                      1711846800,
                      7200,
                      1
                    ],
                    [
                      1729990800,
                      3600,
                      0
                    ],
                    [
                      1743296400,
                      7200,
                      1
                    ],
                    [
                      1761440400,
                      3600,
                      0
                    ],
                    [
                      1774746000,
                      7200,
                      1
                    ],
                    [
                      1792890000,
                      3600,
                      0
                    ],
                    [
                      1806195600,
                      7200,
                      1
                    ],
                    [
                      1824944400,
                      3600,
                      0
                    ],
                    [
                      1837645200,
                      7200,
                      1
                    ],
                    [
                      1856394000,
                      3600,
                      0
                    ],
                    [
                      1869094800,
                      7200,
                      1
                    ],
                    [
                      1887843600,
                      3600,
                      0
                    ],
                    [
                      1901149200,
                      7200,
                      1
                    ],
                    [
                      1919293200,
                      3600,
                      0
                    ],
                    [
                      1932598800,
                      7200,
                      1
                    ],
                    [
                      1950742800,
                      3600,
                      0
                    ],
                    [
                      1964048400,
                      7200,
                      1
                    ],
                    [
                      1982797200,
                      3600,
                      0
                    ],
                    [
                      1995498000,
                      7200,
                      1
                    ]
                  ]
                },
                "online": {
                  "status": 1
                }
              },
              "digest": {
                "togglex": [
                  {
                    "channel": 0,
                    "onoff": 1,
                    "lmTime": 1704798367
                  }
                ],
                "triggerx": [],
                "timerx": []
              }
            },
            "payloadVersion": 1,
            "ability": {
              "Appliance.Config.Key": {},
              "Appliance.Config.WifiList": {},
              "Appliance.Config.Wifi": {},
              "Appliance.Config.Trace": {},
              "Appliance.System.All": {},
              "Appliance.System.Hardware": {},
              "Appliance.System.Firmware": {},
              "Appliance.System.Debug": {},
              "Appliance.System.Online": {},
              "Appliance.System.Time": {},
              "Appliance.System.Clock": {},
              "Appliance.System.Ability": {},
              "Appliance.System.Runtime": {},
              "Appliance.System.Report": {},
              "Appliance.System.Position": {},
              "Appliance.System.DNDMode": {},
              "Appliance.Control.Multiple": {
                "maxCmdNum": 5
              },
              "Appliance.Control.ToggleX": {},
              "Appliance.Control.TimerX": {
                "sunOffsetSupport": 1
              },
              "Appliance.Control.TriggerX": {},
              "Appliance.Control.Bind": {},
              "Appliance.Control.Unbind": {},
              "Appliance.Control.Upgrade": {},
              "Appliance.Control.ConsumptionX": {},
              "Appliance.Control.Electricity": {},
              "Appliance.Control.ConsumptionConfig": {},
              "Appliance.Digest.TriggerX": {},
              "Appliance.Digest.TimerX": {}
            }
          },
          "key": "#whatever#"
        },
        "options": {},
        "pref_disable_new_entities": false,
        "pref_disable_polling": false,
        "source": "integration_discovery",
        "unique_id": "#surely the old UUID: this need to be fixed#",
        "disabled_by": null
}
```
Actually, the 'unique_id' and 'device_id' entries should be permanent and they are never changed through the life of the configuration so you should see these fields as having the old UUID. These are then checked at runtime (in different code paths) to ensure the device who we're talking with is the same original one. By manually changing these you are effectively changing the 'original identity' and things should be fixed.

The bug, previously allowed to update the 'payload' dictionary with actual (polled) device data from a possibly bogus device and so, these might already being reflecting the new device configuration (in this case they're ok and you should leave them)
To be honest, this is the scenario I'm expecting in your configuration (see notes/comments in the reported example). If something doesn't match my expectations then we're in 'unexplored field' and I don't know what could happen.
Anyway, the notes about these sensitive fields should guide you to a succesful manual editing. You should stop HA when manually editing since it could update the full storage file at any time with its own loaded storage and so mess up/invalidate your edits
Make a backup ;)

After editing restart HA and check everything went smooth (crossfingers!)

**Reply by BlackDragon7402** [reply]
Hey,
just did the steps you recommended.
Found the new UUID by searching for the MAC Adress of the new device in the same file.
Replaced the old one at 2 spots (as seen in picture) and restarted. But sady it just created new entities and therefore lost the history.

![grafik](https://github.com/krahabb/meross_lan/assets/119086185/cb76b832-8660-4911-8381-5fb53d1da64b)

If I made something wrong lmk but I guess its not possible since it seems to be tied to the UUID?
Otherwise I will just let it be (made a backup before obivously), losing the history of such a device is not too bad.

**Reply by krahabb** [reply]
Awww..yeah..I've completely forgot that!
Entities are tightly coupled to the device_id (their 'unique_id' which is used to identify them in the entity registry contains the device uuid) and so, when loading the patched config entry it is going to register completely different entities.

If you want to give it another shot, you coul enter the entity registry (always in .storage) and look for the unique_ids containing the old device uuid and patch those too...

But I don't know if this could be definitive :(
...there are so many related things...

likely the device registry too would need to be patched

**Reply by BlackDragon7402** [reply]
Thanks for your help but in this case I will just let it be as is since the gain (historical data) is not worth the effor/risk in my opinion

**Reply by krahabb** [reply]
Yeah..that's why the device UUID mismatch is (somewhat) dangerous and I've preferred avoiding/blocking this scenario. Of course loosing the history is a bit of a shame for your case where you just substituted the device

**Reply by BlackDragon7402** [reply]
No worries
I definitely welcome the change and think its a big improvement to the integration regarding stability/reliability.
Thanks for that and all of your other work :D
